### PR TITLE
Enhance permission denied exception logs

### LINKED ISSUE #874: Clean flank not authorized error messages
**Describe the bug**
> A clear and concise description of what the bug is.

`flank android run` and `flank ios run` produce large error messages that aren't user friendly when the service account isn't authenticated correctly. Let's clean up the error messages to improve the user experience.

**To Reproduce**
> Steps to reproduce the behavior:

1. Update flank.yml to point to a project that flank isn't authorized for
2. Run `flank android run`

**Expected behavior**
> A clear and concise description of what you expected to happen.

A clean error message, something like this:

```
$ flank android run
version: v20.06.2
revision: 08de9c3f4f8ee1ecc3684ecb06391b8521a08f4d

Flank encountered a 403 error when running on project foo-bar. Please verify this credential is authorized for the project.

Caused by: com.google.api.client.googleapis.json.GoogleJsonResponseException: 403 Forbidden
{
  "code" : 403,
  "errors" : [ {
    "domain" : "global",
    "message" : "The caller does not have permission",
    "reason" : "forbidden"
  } ],
  "message" : "The caller does not have permission",
  "status" : "PERMISSION_DENIED"
}
```

**Details (please complete the following information):**
> Have you tested on the latest Flank snapshot?
>
> Post the output of ` flank --version`.

Yes

**Additional context**
> Add any other context about the problem here.

Here's the current error message:

```
flank android run
version: v20.06.2
revision: 08de9c3f4f8ee1ecc3684ecb06391b8521a08f4d
java.io.IOException: Request failed
	at ftl.http.ExecuteWithRetryKt$executeWithRetry$$inlined$withRetry$1.invokeSuspend(ExecuteWithRetry.kt:36)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.ResumeModeKt.resumeMode(ResumeMode.kt:67)
	at kotlinx.coroutines.DispatchedKt.resume(Dispatched.kt:319)
	at kotlinx.coroutines.DispatchedKt.dispatch(Dispatched.kt:301)
	at kotlinx.coroutines.CancellableContinuationImpl.dispatchResume(CancellableContinuationImpl.kt:250)
	at kotlinx.coroutines.CancellableContinuationImpl.resumeImpl(CancellableContinuationImpl.kt:260)
	at kotlinx.coroutines.CancellableContinuationImpl.resumeUndispatched(CancellableContinuationImpl.kt:332)
	at kotlinx.coroutines.EventLoopImplBase$DelayedResumeTask.run(EventLoop.common.kt:483)
	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:270)
	at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:79)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:54)
	at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:36)
	at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
	at ftl.http.ExecuteWithRetryKt.executeWithRetry(ExecuteWithRetry.kt:41)
	at ftl.gc.GcToolResults.getDefaultBucket(GcToolResults.kt:118)
	at ftl.args.ArgsHelper.createGcsBucket(ArgsHelper.kt:124)
	at ftl.args.CreateCommonArgsKt.createCommonArgs(CreateCommonArgs.kt:15)
	at ftl.args.CreateAndroidArgsKt.createAndroidArgs$default(CreateAndroidArgs.kt:12)
	at ftl.args.AndroidArgsCompanion.load$flank(AndroidArgsCompanion.kt:35)
	at ftl.args.AndroidArgsCompanion.load(AndroidArgsCompanion.kt:31)
	at ftl.cli.firebase.test.android.AndroidRunCommand.run(AndroidRunCommand.kt:46)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1769)
	at picocli.CommandLine.access$900(CommandLine.java:145)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2150)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2144)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2108)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:1975)
	at picocli.CommandLine.execute(CommandLine.java:1904)
	at ftl.Main$Companion$main$1.invoke(Main.kt:53)
	at ftl.Main$Companion$main$1.invoke(Main.kt:43)
	at ftl.util.Utils.withGlobalExceptionHandling(Utils.kt:130)
	at ftl.Main$Companion.main(Main.kt:49)
	at ftl.Main.main(Main.kt)
Caused by: com.google.api.client.googleapis.json.GoogleJsonResponseException: 403 Forbidden
{
  "code" : 403,
  "errors" : [ {
    "domain" : "global",
    "message" : "The caller does not have permission",
    "reason" : "forbidden"
  } ],
  "message" : "The caller does not have permission",
  "status" : "PERMISSION_DENIED"
}
	at com.google.api.client.googleapis.json.GoogleJsonResponseException.from(GoogleJsonResponseException.java:150)
	at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:113)
	at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:40)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest$1.interceptResponse(AbstractGoogleClientRequest.java:443)
	at com.google.api.client.http.HttpRequest.execute(HttpRequest.java:1092)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:541)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:474)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.execute(AbstractGoogleClientRequest.java:591)
	at ftl.http.ExecuteWithRetryKt$executeWithRetry$$inlined$withRetry$1.invokeSuspend(ExecuteWithRetry.kt:41)
	... 34 more
```

### PR DESCRIPTION
**Pull request by pawelpasterz**
Fixes #874

### 1. User don't have permission to project (403)

When user don't have permission to project Flank should returns message like:

```

Flank encountered a 403 error when running on project $project_name. Please verify this credential is authorized for the project.
Consider authentication with a Service Account https://github.com/Flank/flank#authenticate-with-a-service-account
or with a Google account https://github.com/Flank/flank#authenticate-with-a-google-account

Caused by: com.google.api.client.googleapis.json.GoogleJsonResponseException: 403 Forbidden
{
  "code" : 403,
  "errors" : [ {
    "domain" : "global",
    "message" : "The caller does not have permission",
    "reason" : "forbidden"
  } ],
  "message" : "The caller does not have permission",
  "status" : "PERMISSION_DENIED"
}

```

### 2. Project not found (404)

When project not found on firebase Flank should return message like:

```

Flank was unable to find project $project_name. Please verify the project id.

Caused by: com.google.api.client.googleapis.json.GoogleJsonResponseException: 404 Not Found
{
  "code" : 404,
  "errors" : [ {
    "domain" : "global",
    "message" : "Project not found: $project_name",
    "reason" : "notFound"
  } ],
  "message" : "Project not found: $project_name",
  "status" : "NOT_FOUND"
}

```

## Checklist

- [X] Documented
- [X] Unit tested
- [X] release_notes.md updated

### CODE REVIEW

File: test_runner/src/main/kotlin/ftl/gc/GcToolResults.kt [hunk 1]
```diff
@@ -114,8 +118,27 @@ object GcToolResults {
             ).executeWithRetry()
     }
 
-    fun getDefaultBucket(projectId: String): String? {
-        val response = service.Projects().initializeSettings(projectId).executeWithRetry()
-        return response.defaultBucket
+    fun getDefaultBucket(projectId: String): String? = try {
+        service.Projects().initializeSettings(projectId).executeWithRetry().defaultBucket
+    } catch (ftlProjectError: FTLProjectError) {
+        // flank need to rewrap exception with additional info about project
```
Comment (bootstraponline) [inline test_runner/src/main/kotlin/ftl/gc/GcToolResults.kt:124]:
```suggestion
        // flank needs to rewrap the exception with additional info about project
```

File: test_runner/src/main/kotlin/ftl/gc/GcToolResults.kt [hunk 2]
```diff
@@ -114,8 +118,27 @@ object GcToolResults {
...
+        when (ftlProjectError) {
+            is PermissionDenied -> throw FlankCommonException(permissionDeniedErrorMessage(projectId, ftlProjectError.message))
+            is ProjectNotFound -> throw FlankCommonException(projectNotFoundErrorMessage(projectId, ftlProjectError.message))
+        }
     }
 }
+
+private val permissionDeniedErrorMessage = { projectId: String, message: String? ->
+    """Flank encountered a 403 error when running on project $projectId. Please verify this credential is authorized for the project.
+Consider authentication with Service Account https://github.com/Flank/flank#authenticate-with-a-service-account
```
Comment (bootstraponline) [inline test_runner/src/main/kotlin/ftl/gc/GcToolResults.kt:134]:
```suggestion
Consider authentication with a Service Account https://github.com/Flank/flank#authenticate-with-a-service-account
```

File: test_runner/src/main/kotlin/ftl/gc/GcToolResults.kt [hunk 3]
```diff
@@ -114,8 +118,27 @@ object GcToolResults {
...
+or with Google account https://github.com/Flank/flank#authenticate-with-a-google-account
```
Comment (bootstraponline) [inline test_runner/src/main/kotlin/ftl/gc/GcToolResults.kt:135]:
```suggestion
or with a Google account https://github.com/Flank/flank#authenticate-with-a-google-account
```

File: test_runner/src/main/kotlin/ftl/gc/GcToolResults.kt [hunk 4]
```diff
@@ -114,8 +118,27 @@ object GcToolResults {
...
+
+$message""".trimIndent()
+}
+
+private val projectNotFoundErrorMessage = { projectId: String, message: String? ->
+    """Flank was unable to find project $projectId. Please verify project id.
```
Comment (bootstraponline) [inline test_runner/src/main/kotlin/ftl/gc/GcToolResults.kt:141]:
```suggestion
    """Flank was unable to find project $projectId. Please verify the project id.
```

**Comment by bootstraponline** [comment]
The pull request looks great! I left some small comments about updating the wording.

File: test_runner/src/main/kotlin/ftl/http/ExecuteWithRetry.kt [hunk 5]
```diff
@@ -2,44 +2,60 @@ package ftl.http
...
+} catch (ioError: IOException) {
+    ExecutionBlockResult(null, ioError)
+}
 
-    throw IOException("Request failed", lastErr)
+private data class ExecutionBlockResult<T>(private val success: T?, private val error: IOException?) {
+    fun isSuccess() = success != null
+    fun success() = success!!
+    fun error() = error!!
 }
```
Comment (jan-gogo) [inline test_runner/src/main/kotlin/ftl/http/ExecuteWithRetry.kt:47]: This commit brings 16 lines of unnecessary abstraction. Code does nothing more then wrapping `try catch`, but is much harder to understand. Pls revert it. The only one reasonable change which I see here is removing redundant`return@runBlocking` from `return@runBlocking block()`.
BTW kotlin already have functional abstraction for `try catch` it's `runCatching`, but it doesn't bring much more to error handling IMO is usefully only for specific cases.

Comment (pawelpasterz) [reply test_runner/src/main/kotlin/ftl/http/ExecuteWithRetry.kt:47]: I think @jan-gogo is right, we should revert this change and maybe think about different approach

Comment (adamfilipow92) [reply test_runner/src/main/kotlin/ftl/http/ExecuteWithRetry.kt]: Thanks for feedback! I revert changes

File: docs/permissions_denied_behavior.md [hunk 6]
```diff
@@ -28,6 +28,8 @@ Caused by: com.google.api.client.googleapis.json.GoogleJsonResponseException: 40
 
 ```
 
+You can reproduce it by set PROJECT_ID to project where your current firebase account doesn't have permission.
```
Comment (bootstraponline) [inline docs/permissions_denied_behavior.md:31]:
```suggestion
You can reproduce the error by setting PROJECT_ID to a project that the firebase account doesn't have permission to access.
```

File: docs/permissions_denied_behavior.md [hunk 7]
```diff
@@ -50,4 +52,6 @@ Caused by: com.google.api.client.googleapis.json.GoogleJsonResponseException: 40
 
 ```
 
+You can reproduce it by set PROJECT_ID to not existing project.
```
Comment (bootstraponline) [inline docs/permissions_denied_behavior.md:55]:
```suggestion
You can reproduce the error by setting PROJECT_ID to a project that doesn't exist.
```

File: test_runner/src/main/kotlin/ftl/gc/GcToolResults.kt [hunk 8]
```diff
@@ -114,8 +118,27 @@ object GcToolResults {
...
+        // flank needs to rewrap the exception with additional info about project
+        when (ftlProjectError) {
+            is PermissionDenied -> throw FlankCommonException(permissionDeniedErrorMessage(projectId, ftlProjectError.message))
+            is ProjectNotFound -> throw FlankCommonException(projectNotFoundErrorMessage(projectId, ftlProjectError.message))
+        }
     }
 }
+
+private val permissionDeniedErrorMessage = { projectId: String, message: String? ->
+    """Flank encountered a 403 error when running on project $projectId. Please verify this credential is authorized for the project.
```
Comment (jan-gogo) [inline test_runner/src/main/kotlin/ftl/gc/GcToolResults.kt:133]:
```suggestion
    """Flank encountered a 403 error when running on project $projectId. Please verify this credential is authorized for the project and have required permissions.
```

File: test_runner/src/test/kotlin/ftl/gc/GcToolResultsTest.kt [hunk 9]
```diff
@@ -1,43 +1,157 @@
...
+            .setProjectId(projectId)
             .setHistoryId("mockId")
 
         assertThat(GcToolResults.createToolResultsHistory(args)).isEqualTo(expected)
     }
+
+    @Test
+    fun `getDefaultBucket on 403 error should throw exception with specific message`() {
+        val expected = """
+            Flank encountered a 403 error when running on project $projectId. Please verify this credential is authorized for the project.
```
Comment (jan-gogo) [inline test_runner/src/test/kotlin/ftl/gc/GcToolResultsTest.kt:59]:
```suggestion
            Flank encountered a 403 error when running on project $projectId. Please verify this credential is authorized for the project and have required permissions.
```

Comment (bootstraponline) [reply test_runner/src/test/kotlin/ftl/gc/GcToolResultsTest.kt:59]: ` and has the required permissions.`

File: test_runner/src/main/kotlin/ftl/gc/GcToolResults.kt (hunk 8 above)
Comment (bootstraponline) [reply test_runner/src/main/kotlin/ftl/gc/GcToolResults.kt:133]: `and has the required permissions.`

**Pull request merged**